blocks/graph.py

```diff
@@ -205,18 +204,24 @@ def replace(self, replacements):
         replacement_keys_cur = []
         replacement_vals_cur = []
         # Sort `replacements` in topological order
-        for node in apply_nodes_sorted:
-            for input_ in node.inputs:
-                if input_ in replacements:
-                    if input_ not in replacement_keys_cur:
-                        replacement_keys_cur.append(input_)
-                        replacement_vals_cur.append(replacements[input_])
-
-        # Add outputs of the computation graph
-        for output in self.outputs:
-            if output in replacements:
-                replacement_keys_cur.append(output)
-                replacement_vals_cur.append(replacements[output])
+        # variables in self.variables are in topological order
+        remaining_replacements = replacements.copy()
+        for variable in self.variables:
+            if variable in replacements:
+                if has_roles(variable, [AUXILIARY]):
+                    warnings.warn(
+                        "replace method was asked to replace a variable ({}) "
+                        "that is an auxiliary variable.".format(variable))
+                replacement_keys_cur.append(variable)
```

Comment:
Well, self.variables should not contain duplicates thus remaining_replacements.pop(variable) should not fail, otherwise it means the variable was already seen and popped out.